Validation/RecoEgamma/plugins/ElectronMcMiniAODSignalValidator.cc

```diff
@@ -0,0 +1,392 @@
+            if ( el1->charge() * el2->charge() < 0. )
+            {
+                h1_ele_mee_os->Fill(sqrt(mee2));
+            }
+        }
+    }
+
+    //===============================================
+    // charge mis-ID
+    //===============================================
+
+    int mcNum=0, gamNum=0, eleNum=0 ;
+//    bool matchingID;//, matchingMotherID ;
+    bool matchingMotherID ;
+
+    //===============================================
+    // association mc-reco
+    //===============================================
+
+    for(size_t i=0; i<genParticles->size(); i++) {  
+        // number of mc particles
+        mcNum++ ;
+
+        // counts photons
+        if ( (*genParticles)[i].pdgId() == 22 )
+            { gamNum++ ; }
+
+        // select requested mother matching gen particle
+        // always include single particle with no mother
+        const Candidate * mother = (*genParticles)[i].mother(0) ;
+        matchingMotherID = false ;
+        for ( unsigned int ii=0 ; ii<matchingMotherIDs_.size() ; ii++ ) {
+            if ( (mother == 0) || ((mother != 0) &&  mother->pdgId() == matchingMotherIDs_[ii]) )
+                { matchingMotherID = true ; 
+//			std::cout << "matchingMotherID :" << matchingMotherIDs_[ii] << std::endl ;
+            }
+        }
+        if (!matchingMotherID) continue ;
+
+        // electron preselection
+        if ((*genParticles)[i].pt()> maxPt_ || std::abs((*genParticles)[i].eta())> maxAbsEta_)
+            { continue ; }
+        eleNum++;
+
+        // find best matched electron
+        bool okGsfFound = false ;
+        bool passMiniAODSelection = true ;
+        double gsfOkRatio = 999999. ;
+        pat::Electron bestGsfElectron ;
+        for (const pat::Electron &el : *electrons ) {
+            double dphi = el.phi()-(*genParticles)[i].phi() ;
+            if (std::abs(dphi)>CLHEP::pi)
+                { dphi = dphi < 0? (CLHEP::twopi) + dphi : dphi - CLHEP::twopi ; }
+            double deltaR2 = (el.eta()-(*genParticles)[i].eta()) * (el.eta()-(*genParticles)[i].eta()) + dphi * dphi;
+            if ( deltaR2 < deltaR2_ )
+            {
+                if ( ( ((*genParticles)[i].pdgId() == 11) && (el.charge() < 0.) ) ||
+                ( ((*genParticles)[i].pdgId() == -11) && (el.charge() > 0.) ) )
+                {
+                    double tmpGsfRatio = el.p()/(*genParticles)[i].p() ;
+                    if ( std::abs(tmpGsfRatio-1) < std::abs(gsfOkRatio-1) )
+                        {
+                        gsfOkRatio = tmpGsfRatio;
+                        bestGsfElectron=el;
+                        okGsfFound = true;
+                    }
+                }
+            }
+        }
+     
+        if (! okGsfFound) continue ;
+
+        //------------------------------------
+        // analysis when the mc track is found
+        //------------------------------------
+        passMiniAODSelection = bestGsfElectron.pt() >= 5.;
```

Comment:
this will be hard to maintain - is it not possible to just use the miniAOD collections instead?
@archiron @deguio